apollo-router/src/http_ext.rs

```diff
@@ -12,119 +12,207 @@ use std::ops::DerefMut;
+        }
     }
 }
 
-impl Hash for IntoHeaderValue {
-    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        self.as_bytes().hash(state);
+impl From<&'_ str> for TryIntoHeaderName {
+    fn from(value: &'_ str) -> Self {
+        Self {
+            result: value.try_into(),
+        }
     }
 }
 
-impl IntoHeaderName {
-    fn as_bytes(&self) -> &[u8] {
-        match self {
-            IntoHeaderName::String(s) => s.as_bytes(),
-            IntoHeaderName::HeaderName(h) => h.as_str().as_bytes(),
+impl From<Vec<u8>> for TryIntoHeaderName {
+    fn from(value: Vec<u8>) -> Self {
+        Self {
+            result: value.try_into(),
         }
     }
 }
 
-impl IntoHeaderValue {
-    fn as_bytes(&self) -> &[u8] {
-        match self {
-            IntoHeaderValue::String(s) => s.as_bytes(),
-            IntoHeaderValue::HeaderValue(v) => v.as_bytes(),
+impl From<String> for TryIntoHeaderName {
+    fn from(value: String) -> Self {
+        Self {
+            result: value.try_into(),
         }
     }
 }
 
-impl<T> From<T> for IntoHeaderName
-where
-    T: std::fmt::Display,
-{
-    fn from(name: T) -> Self {
-        IntoHeaderName::String(name.to_string())
+impl From<header::HeaderValue> for TryIntoHeaderValue {
+    fn from(value: header::HeaderValue) -> Self {
+        Self { result: Ok(value) }
     }
 }
 
-impl<T> From<T> for IntoHeaderValue
-where
-    T: std::fmt::Display,
-{
-    fn from(name: T) -> Self {
-        IntoHeaderValue::String(name.to_string())
+impl From<&'_ header::HeaderValue> for TryIntoHeaderValue {
+    fn from(value: &'_ header::HeaderValue) -> Self {
+        Self {
+            result: Ok(value.clone()),
+        }
     }
 }
 
-impl TryFrom<IntoHeaderName> for HeaderName {
-    type Error = http::Error;
+impl From<&'_ [u8]> for TryIntoHeaderValue {
+    fn from(value: &'_ [u8]) -> Self {
+        Self {
+            result: value.try_into(),
+        }
+    }
+}
 
-    fn try_from(value: IntoHeaderName) -> Result<Self, Self::Error> {
-        Ok(match value {
-            IntoHeaderName::String(name) => HeaderName::try_from(name)?,
-            IntoHeaderName::HeaderName(name) => name,
-        })
+impl From<&'_ str> for TryIntoHeaderValue {
+    fn from(value: &'_ str) -> Self {
+        Self {
+            result: value.try_into(),
+        }
+    }
+}
+
+impl From<Vec<u8>> for TryIntoHeaderValue {
+    fn from(value: Vec<u8>) -> Self {
+        Self {
+            result: value.try_into(),
+        }
+    }
+}
+
+impl From<String> for TryIntoHeaderValue {
+    fn from(value: String) -> Self {
+        Self {
+            result: value.try_into(),
+        }
     }
 }
 
-impl TryFrom<IntoHeaderValue> for HeaderValue {
-    type Error = http::Error;
+impl Eq for TryIntoHeaderName {}
+
+impl PartialEq for TryIntoHeaderName {
+    fn eq(&self, other: &Self) -> bool {
+        match (&self.result, &other.result) {
+            (Ok(a), Ok(b)) => a == b,
+            (Err(_), Err(_)) => true,
```

Comment:
Why are two invalid headers equal?

What you wrote here looks like an Eq implementation, PartialEq being `(Err(_), Err(_)) => false,`?

Reply:
We use this as keys in a `MultiMap`, which wraps `std::collections::HashMap` which [documents](https://doc.rust-lang.org/std/collections/struct.HashMap.html):

> It is required that the keys implement the [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html) and [Hash](https://doc.rust-lang.org/std/hash/trait.Hash.html) traits

When there’s any error, `header_map` ends up returning an error and not a map so map key equality does not matter.